### PR TITLE
Make kafka-cluster chunk save messages partition aware.

### PR DESCRIPTION
**Pull request by woodsaj**
- Move partition fetching/validation to shared kafka lib
- Refactor sending MetricPersist messages to set the correct
  PartitionKey.
- Update consumer to only consume from configured partitions

Issue #452

### CODE REVIEW

File: mdata/notifierKafka/notifierKafka.go [hunk 1]
```diff
@@ -176,36 +173,57 @@ func (c *NotifierKafka) flush() {
...
 	}
 
-	msg := mdata.PersistMessageBatch{Instance: c.instance, SavedChunks: c.buf}
-	c.buf = nil
-
-	go func() {
-		log.Debug("kafka-notifier sending %d batch metricPersist messages", len(msg.SavedChunks))
-
-		data, err := json.Marshal(&msg)
+	// In order to correctly routes the saveMessages to the correct partition,
```
Comment (replay) [inline mdata/notifierKafka/notifierKafka.go]: `route` without `s`

File: mdata/notifierKafka/notifierKafka.go [hunk 2]
```diff
@@ -176,36 +173,57 @@ func (c *NotifierKafka) flush() {
...
-			log.Fatal(4, "kafka-notifier failed to marshal persistMessage to json.")
+			log.Error(3, "kafka-cluster: failed to lookup metricDef with id %s", msg.Key)
+			continue
 		}
-		buf := new(bytes.Buffer)
+		buf := bytes.NewBuffer(c.bPool.Get())
 		binary.Write(buf, binary.LittleEndian, uint8(mdata.PersistMessageBatchV1))
-		buf.Write(data)
+		encoder := json.NewEncoder(buf)
+		pMsg = mdata.PersistMessageBatch{Instance: c.instance, SavedChunks: c.buf[i:i]}
```
Comment (replay) [inline mdata/notifierKafka/notifierKafka.go]: i'm confused about `[i:i]`, wouldn't that always return an empty list? should that be `[i:i+1]`?

File: kafka/partitions.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	for i, topic := range topics {
+		partitions, err = client.Partitions(topic)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get partitions for topic %s. %s", topic, err)
+		}
+		if len(partitions) == 0 {
+			return nil, fmt.Errorf("No partitions returned for topic %s", topic)
+		}
+		if i > 0 {
+			if len(partitions) != partitionCount {
```
Comment (woodsaj) [inline kafka/partitions.go]: Kafka uses sequential partition numbers.  So if the partitionCount is the same, then the partitions will also be the same.

File: mdata/notifierKafka/notifierKafka.go (hunk 2 above)
Comment (woodsaj) [inline mdata/notifierKafka/notifierKafka.go]: Yep, that should definitely be `[i:i+1]`

File: input/kafkamdm/kafkamdm.go [hunk 4]
```diff
@@ -139,51 +139,23 @@ func ConfigProcess(instance string) {
...
-				log.Fatal(4, "kafka-mdm: configured partitions not in list of available partitions. missing %v", missing)
-			}
+	availParts, err := kafka.GetPartitions(client, topics)
+	if err != nil {
+		log.Fatal(4, "kafka-mdm: %s", err.Error())
+	}
+	log.Info("kafka-mdm: available partitions %v", availParts)
+	if partitionStr == "*" {
+		partitions = availParts
+	} else {
```
Comment (woodsaj) [inline input/kafkamdm/kafkamdm.go]: good call. i have made that change.

**Pull request merged**